Remove dependency to spree_api - step 2 - routes and views

luisramos0 opened the pull request:
#### What? Why?

Closes #4054 and #4060 and #4069

This PR finished up what is required to remove spree_api dependency.
The best and most important change here is: "Merge Spree::Api::BaseController with Api::BaseController. All api controllers inherit from Api::BaseController now."

#### What should we test?
This is a very big change, this is a basic verification plan I can think of (there's nothing specific to any of these views being changed except what was discussed in #4111):
- Admin - Config - Taxonomies - Edit taxonomies and use them when creating and editing products
- Admin - Enterprises - Shop Preferences - Category Ordering (make sure the list of taxons is correct)
- Admin - Products - Bulk update products and bulk edit variants
- Admin - Products - Category filter and Category column (view and edit)
- Admin - Products - Upload product image in the bulk product edit page
- Admin - Products - Inventory - update a variant override stock level
- Admin - OC - create and edit order cycles
- Shopfront - change OC in the frontoffice and check that the list of products renders correctly and the filters/search works
- Shopfront - verify the checkout works in particular the entry of addresses
- Admin - Orders - New order - create an order, add variants both normal variants as well as variants with overridden prices and stock and check they are picked up correctly (test the input to search for variants)
- Admin - Orders - Edit page - add, edit quantity and remove variants from the order. Ship the order.

#### Release notes
Changelog Category: Changed
Removed spree_api dependency, OFN API is now self sufficient and can now grow as required using a specific tech stack.

#### Dependencies

This PR goes on top of quite a few PRs that make up the base branch: remove_spree_api_2_base
- #4112 Convert spree/api/shipments views from rabl to ams 
- #4104 Convert spree/api/taxons views from rabl to AMS 
- #4101 Convert spree/api/products and spree/api/variants views from rabl to AMS 
- #4068 Convert order_cycle.rabl to Api::OrderCycleSerializer 
- #4073 Replace update_prod_image.rabl with ImageSerializer in api/product_image_controller 
- #4064 Replace views/admin/variants/search.rabl with Api::Admin::VariantSerializer 
- #4111 Delete now irrelevant authorize_api endpoint and logic (this one is now incorporated in this PR, see [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/4111#issuecomment-521614215) for details)
- #4063 Remove dead code under views/spree/shared 

This PR can only be tested when all PRs above are merged.

luisramos0 commented:
unblocked :tada: I just need to resolved conflicts

sauloperez commented on app/assets/javascripts/admin/bulk_product_update.js.coffee:
```diff
@@ -39,12 +39,7 @@ angular.module("ofn.admin").controller "AdminProductEditCtrl", ($scope, $timeout
   $scope.DisplayProperties = DisplayProperties
 
   $scope.initialise = ->
-    SpreeApiAuth.authorise()
```
how does that happen now?

luisramos0 replied:
this is explained in #4111 [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/4111#issuecomment-517661713)

sauloperez commented on app/controllers/api/base_controller.rb:
```diff
@@ -1,16 +1,44 @@
+    ssl_allowed
+
+    # Include these because we inherit from ActionController::Metal
+    #   rather than ActionController::Base and these are required for AMS
```
I'd be great to understand why and document it, maybe something for a new spike issue.

luisramos0 replied:
these lines of code were already there and this new comment is basically an improvement from the removed top class comment "# Includes the minimum machinery required by ActiveModelSerializers"

I dont know exactly why they are needed but it looks like they are, as mentioned, "the minimum machinery required by ActiveModelSerializers"

Do you think this is not needed?

luisramos0 replied:
yes, Metal is a stripped down version for apis, 40% performance improvement in some cases according to this https://stackoverflow.com/questions/18024337/what-actually-does-rails-actioncontrollermetal

sauloperez commented on app/controllers/api/base_controller.rb:
```diff
@@ -19,16 +47,62 @@ def respond_with_conflict(json_hash)
     def authenticate_user
-      @current_api_user = try_spree_current_user
-      super
+      return if @current_api_user = try_spree_current_user
```
Is this really equivalent? previously, we called `super`, which means that the parent's logic was always executed. Is the logic below not a copy&paste from that?

kristinalim replied:
It looks equivalent logically. :+1: When executing `super`, the method is exited if `@current_api_user` is truthy, which is what happens here too.

sauloperez commented on lib/spree/api/controller_setup.rb:
```diff
@@ -24,7 +22,6 @@ def self.included(klass)
           prepend_view_path Rails.root + "app/views"
           append_view_path File.expand_path("../../../app/views", File.dirname(__FILE__))
 
-          self.responder = Spree::Api::Responders::AppResponder
```
where the responders now?

luisramos0 replied:
I dont think we have responders now (we dont use them for sure), I think we still get the rails default.
Anyway, in the API space, do we need them? it's only json in the api.

sauloperez replied:
what about the endpoints we haven't implemented ourselves? Don't they rely on Spree's responders?

luisramos0 replied:
We are dropping spree_api altogether, all we have from now on are the endpoints we implement in OFN and all OFN api endpoints are already using AMS (that was the work done in all the previous related PRs).
New endpoints in OFN can be inspired/copied from spree_api but need to be converted to AMS. We are done with rabl :tada:

luisramos0 commented:
just realized config/routes/spree.rb still had api.users.authorise_api route, deleted it in existing commit.

luisramos0 commented:
staging this in katuma so I can use it to support the swagger work I am doing on #4254

luisramos0 commented:
testing this PR a bit in staging, I found that editing a user in the backoffice was not showing the api key part of the form. This part of the form was in spree_api gem as a partial injected through a deface override.
I now added this partial to OFN in this PR and added a feature spec to test it's basic usage "generating an api key for a user".

daniellemoorhead commented:
Hey @luisramos0 I used Katuma staging for testing today but have redeployed this one on it so that you can continue testing when you wake up :)

lin-d-hop commented:
Added blocked label.

This PR cannot be merged until  /api/orders/${order_number}.json is added back in. This is now a key requirement for a number of UK users' accounting package integrations via Zapier.

Thanks!

luisramos0 commented:
👍 I have created the issue to add the api/order/{order_number} endpoint here #4264 

So, this PR should only be tested after a PR that fixes #4264

luisramos0 commented:
rebased to resolve conflicts.

luisramos0 commented:
#4264 is not resolved, this PR is not blocked anymore.

No code changes since the approvals, just two rebases, so we are good for testing this one now 🎉

RachL commented:
@luisramos0 can I stage this or do you need to resolve the conflicts first?

luisramos0 commented:
better if I rebase now. done :-)

RachL commented:
@luisramos0 the fact that the build is failing should not be a concern?

sigmundpetersen commented:
Build failures seems consistent

luisramos0 commented:
:+1: yah, blocked until I fix the build. I dont understand it yet.

luisramos0 commented:
The build is green now, I will request a quick code review on these two last commits that fix the specs.

luisramos0 commented:
ok, thanks Pau, let's say this is ready for testing! 💪

RachL commented:
Ok sorry that it has tacken me a so long time to test this. I believe this is ready to go!

Testing notes: https://docs.google.com/document/d/1BfGPBvBD3bPmcb5CUwerx4fHjmRy8q8gKIyiPWdWslI/edit#

luisramos0 commented:
OMG, we did it! Thanks everyone for your support!!!

sauloperez commented:
OMG we did it? http://giphygifs.s3.amazonaws.com/media/cLLkqzdxu9Rgk/giphy.gif
